### PR TITLE
Modernize Gradle build

### PR DESCRIPTION
**Pull request by melix**
This PR implements a number of improvements in the Gradle build.
It's largely a code cleanup, making sure that the build scripts of each project are declarative only. All build logic has been moved in its own subproject.

It was quite hard to follow what you've done before, but I _think_ I managed to reproduce the previous behavior, in particular with where the artifacts are built (outside of the normal `build` directories).

I have migrated everything to the lazy APIs and I also made use of the new Maven Publish plugin. A side effect is that the `core` project now publishes javadocs and sources. By the way, the javadocs are full of errors which would need to be fixed.

I don't know how to produce docs for the Android projects, since I'm not an expert, but there was a lot of commented out code in the build that I just removed: for external folks like me it doesn't really make sense to have commented out code: either the code is useful in which case it shouldn't be commented out, but potentially activated by some flags, or it should be removed.

Here's a build scan after calling `dist` with this PR: https://scans.gradle.com/s/lst5evsi2rr3k

And the resulting output directory layout:

image

Please make sure to review the generated artifacts as I may have made some mistakes in the transition. I'm happy to answer questions on the new build layout if needed.

### CODE REVIEW

File: constraintlayout/core/build.gradle [hunk 1]
```diff
@@ -1,50 +1,14 @@
...
+plugins {
+    id 'androidx.build.java-library'
+    id 'androidx.build.publishing'
+}
 
-compileJava {
-    targetCompatibility = 1.8
-    sourceCompatibility = 1.8
+globalConfig {
+    pomName = 'Android ConstraintLayout Solver'
```
Comment (camaelon) [inline constraintlayout/core/build.gradle:7]: Android ConstraintLayout Core

File: constraintlayout/core/build.gradle [hunk 2]
```diff
@@ -1,50 +1,14 @@
...
+    pomDescription = 'Solver for ConstraintLayout'
```
Comment (camaelon) [inline constraintlayout/core/build.gradle:8]: ConstraintLayout core library

**Review by camaelon — COMMENTED** [review]
Thanks a lot -- I was able to patch it locally and it all seems to work :) -- the test on github fail I believe due to the jcenter resolution issues. Note that the old structure you had to replicate ... is not really necessary. This is a remnant of when we were publishing the artifacts via the Android Studio SDK Manager.

The only thing we need nowadays is to be able to publish to maven (and to mavenLocal for local tests/development) -- this might help simplify this CL.

**Comment by camaelon** [comment]
Oh and for javadoc -- not surprised to see some errors, I don't think we ever ran it on constraintlayout-core. The one we actually need are for the android libs, but that's another can of worm that's beyond this CL :)

**Comment by melix** [comment]
> The only thing we need nowadays is to be able to publish to maven (and to mavenLocal for local tests/development) -- this might help simplify this CL.

Indeed it's much simpler, so I can get rid of all the special setup! If I get you right, you need:

- a way to publish to `mavenLocal()`: done
- a way to publish to a local test repository: done
- a way to publish to an external repository: technically no difference with providing a local URI

If so I can get rid of all the custom logic to put artifacts in different `build` directories. How does that sound?

**Comment by camaelon** [comment]
sounds good to me :)

**Comment by melix** [comment]
Last question: if I remove the custom logic, what is the custom properties file used for? Should I remove it too? If not where should it be put?

**Comment by melix** [comment]
I have removed _all_ custom code, including the special zip file. If it has to be reintroduced just let me know what it's supposed to do and where it should be produced.

**Comment by camaelon** [comment]
Nice! For the zip file, the way it works currently is that we are using it for our internal publication process; the zip contains the maven info+artifacts directly, so gradlew dist generate a zip like that:
core/constraintlayout-core-1.0.0-beta01.jar.md5  
core/constraintlayout-core-1.0.0-beta01.pom.md5  
core/constraintlayout-core-1.0.0-beta01.pom.sha1  
core/source.properties  
core/constraintlayout-core-1.0.0-beta01.jar  
core/constraintlayout-core-1.0.0-beta01.jar.sha1  
core/constraintlayout-core-1.0.0-beta01.pom  

Let me try your latest changes, reading the doc on the publication process the intermediate directory structure doesn't seem to matter for the zip, as long as it contains pom files and files at the same level adhere to maven naming convention. So simply zipping the result of what you now have ought to be enough.

**Comment by melix** [comment]
Ok maybe I wasn't clear. In the previous version I created the zip, with the properties file inside, etc. Then you said that you don't need the custom setup anymore, in particular with the build directory layout. So my question is: should I restore the zip task, and if I do, where do you expect the zip file to be put?

**Comment by camaelon** [comment]
it would be nice to create a zip as it simplifies the process when publishing to maven.google.com, but if that's annoying to do we could skip that. It doesn't really matter where the zips end up to be (right now they are in out/dist); else from the doc the internal directory hierarchy inside the zip doesn't seem to matter either.

**Comment by melix** [comment]
Done, I reintroduced a `dist` task which generates the zips into `build/dist`.

**Comment by camaelon** [comment]
looks good! many thanks!
